factor/grace/classes/factor.php

```diff
@@ -184,4 +203,25 @@ public function possible_states($user) {
             \tool_mfa\plugininfo\factor::STATE_NEUTRAL
         );
     }
+
+    /**
+     * Grace factor implementation.
+     *
+     * If grace period should redirect at end, make this a no-redirect url.
+     *
+     * @return void
+     */
+    public function get_no_redirect_urls() {
+        $redirect = get_config('factor_grace', 'forcesetup');
+
+        if ($this->get_state(true) === \tool_mfa\plugininfo\factor::STATE_NEUTRAL && $redirect) {
```

Comment:
Micro optimisation but I'd put simpler boolean checks first in a multi-conditional